mshadow/cuda/tensor_gpu-inl.cuh

```diff
@@ -488,6 +548,70 @@ inline void AddTakeGrad(Tensor<gpu, 2, DType> dst,
        src.size(0),
        src.size(1));
 }
+
+template<typename IndexType, typename DType>
+inline void AddTakeGradLargeBatch(Tensor<gpu, 2, DType> dst,
+                                  const Tensor<gpu, 1, IndexType>& sorted,
+                                  const Tensor<gpu, 1, IndexType>& index,
+                                  const Tensor<gpu, 2, DType> &src) {
+  const int kWarpBits = kMemUnitBits;
+  const int SZ = 4;
+  const int block_dim_x = 1 << kWarpBits;
+  const int block_dim_y = 4;
+  const int grid_dim_x = (src.size(0) + block_dim_y - 1) / block_dim_y;
+  const int grid_dim_y = (src.size(1) + block_dim_x * SZ - 1) / (block_dim_x * SZ);
+  dim3 dimBlock(block_dim_x, block_dim_y);
+  dim3 dimGrid(grid_dim_x, grid_dim_y);
+
+  CHECK_EQ(dst.size(1), src.size(1)) << "AddTakeGradLargeBatch: shape mismatch";
+  CHECK_EQ(index.size(0), src.size(0)) << "AddTakeGradLargeBatch: shape mismatch";
+  CheckLaunchParam(dimGrid, dimBlock, "AddTakeGradLargeBatch");
+  cudaStream_t stream = Stream<gpu>::GetStream(dst.stream_);
+
+  AddTakeGradLargeBatchKernel<kWarpBits, SZ, DType>
+      <<<dimGrid, dimBlock, 0, stream>>>
+      (dst.dptr_,
+       sorted.dptr_,
+       index.dptr_,
+       src.dptr_,
+       static_cast<int>(src.size(0)),
+       static_cast<int>(src.size(1)));
+}
+
+template<typename KDType, typename VDType>
+inline void SortByKey(Tensor<gpu, 1, KDType> keys, Tensor<gpu, 1, VDType> values,
+                      bool is_ascend) {
+  CHECK_EQ(keys.CheckContiguous(), true);
+  CHECK_EQ(values.CheckContiguous(), true);
+  cudaStream_t stream = Stream<gpu>::GetStream(keys.stream_);
+  thrust::device_ptr<KDType> key_iter = thrust::device_pointer_cast(keys.dptr_);
+  thrust::device_ptr<VDType> value_iter = thrust::device_pointer_cast(values.dptr_);
+  if (is_ascend) {
+    thrust::stable_sort_by_key(
+#if CUDA_VERSION >= 7000
+      thrust::cuda::par.on(stream),
```

Comment:
May need to synchronize here... The streaming system of Thrust seems to be not so reliable.
 
https://thrust.github.io/doc/group__execution__policies.html#ga64da56b76f487725a99818cd63662dac
"Even when using CUDA streams with thrust::cuda::par, there is no guarantee of concurrency. Algorithms which return a data-dependent result or whose implementations require temporary memory allocation may cause blocking synchronization events. Moreover, it may be necessary to explicitly synchronize through cudaStreamSynchronize or similar before any effects induced through algorithm execution are visible to the rest of the system. Finally, it is the responsibility of the caller to own the lifetime of any CUDA streams involved."